_blog/2023-11-15-ssg-in-rust.mdx

```diff
@@ -0,0 +1,355 @@
+```rust
+// src/main.rs
+async fn styles() -> impl IntoResponse {
+    Response::builder()
+        .status(StatusCode::OK)
+        .header("Content-Type", "text/css")
+        .body(include_str!("../templates/styles.css").to_owned())
+        .unwrap()
+}
+```
+
+Then I added it to the HTML output before pushing the HTML to the empty string buffer:
+```rust
+html_output.push_str(r#"<link rel="stylesheet" type="text/css" href="/styles.css">"#);
+```
+
+After starting up local development and quickly checking the `/hello_world` route, I could see that the text was in fact now centre-aligned on the page. Great. Time to move onto the next part.
+
+## 36:28 - Adding Open Graph tags
+This was a part I struggled with quite a lot. I remember when I last logged into Bearblog for my own blog articles that there was a bit at the top indented by three dashes where you could just add properties to an article and it would add the properties in the fenced off text block to OpenGraph meta tags, but I didn't know what it was called. I did some extensive Googling, and realised it was called "frontmatter" - as in the first bit of a book. That makes sense.
+
+I had a quick look for `pulldown-cmark` compatible libraries to see what would turn up. Unfortunately, there was nothing and I lost quite a bit of time trying to figure out what I was supposed to do. I did some more extensive searching and concluded that I needed to parse the text manually - which was when I came across the library `yaml-front-matter`. This had to be it. I used `cargo add yaml-front-matter` and then quickly checked the GitHub repo. Thankfully, there was a great example showing an example that had both the front-matter block as well as some other text below it - meaning I can use it!
+
+I got to work and wrote a function that would comprise of the HTML page head, which would get inserted before the markdown text itself, as well as moving the stylesheet link tag into this function and adding some extra helpful HTML boilerplate:
+```rust
+// src/main.rs
+#[derive(Deserialize)] 
+struct PageMetadata { 
+    title: String, 
+    description: String, 
+} 
+ 
+fn get_page_header(input: &str) -> String { 
+    let document: Document<PageMetadata> = YamlFrontMatter::parse::<PageMetadata>(input).unwrap(); 
+    let PageMetadata { title, description } = document.metadata; 
+ 
+    let mut html_output = String::new(); 
+    html_output.push_str("<head>"); 
+    html_output.push_str( 
+        r#"<meta charset="UTF-8"/> 
+    <meta name="viewport" content="width=device-width, initial-scale=1.0"/> 
+    <meta http-equiv="X-UA-Compatible" content="ie=edge"/> 
+    <link rel="stylesheet" type="text/css" href="/styles.css">"#); 
+    html_output.push_str(&format!( 
+        "<title>{title}</title> 
+        <meta property=\"og:title\" content=\"{title}\"/> 
+        <meta property=\"og:description\" content=\"{description}\"/>" 
```

Comment:
This one could have `r#` as well IIRC

Reply:
I am not sure about this, do `r#` strings allow variable printing?